spras/config/schema.py

```diff
@@ -0,0 +1,168 @@
+- `CaseInsensitiveEnum` (see ./util.py)
+"""
+
+import re
+from typing import Annotated, Optional
+
+from pydantic import AfterValidator, BaseModel, ConfigDict, Field
+
+from spras.config.util import CaseInsensitiveEnum
+
+# Most options here have an `include` property,
+# which is meant to make disabling parts of the configuration easier.
+# When an option does not have a default, it means that it must be set by the user.
+
+class SummaryAnalysis(BaseModel):
+    include: bool
+
+    # We prefer to never allow extra keys, to prevent
+    # any user mistypes.
+    model_config = ConfigDict(extra='forbid')
+
+class CytoscapeAnalysis(BaseModel):
+    include: bool
+
+    model_config = ConfigDict(extra='forbid')
+
+# Note that CaseInsensitiveEnum is not pydantic: pydantic
+# has special support for enums, but we avoid the
+# pydantic-specific "model_config" key here for this reason.
+class MlLinkage(CaseInsensitiveEnum):
+    ward = 'ward'
+    complete = 'complete'
+    average = 'average'
+    single = 'single'
+
+class MlMetric(CaseInsensitiveEnum):
+    euclidean = 'euclidean'
+    manhattan = 'manhattan'
+    cosine = 'cosine'
+
+class MlAnalysis(BaseModel):
+    include: bool
+    aggregate_per_algorithm: bool = False
+    components: int = 2
+    labels: bool = True
+    kde: bool = False
+    remove_empty_pathways: bool = False
+    linkage: MlLinkage = MlLinkage.ward
+    metric: MlMetric = MlMetric.euclidean
+
+    model_config = ConfigDict(extra='forbid')
+
+class EvaluationAnalysis(BaseModel):
+    include: bool
+    aggregate_per_algorithm: bool = False
+
+    model_config = ConfigDict(extra='forbid')
+
+class Analysis(BaseModel):
+    summary: SummaryAnalysis = SummaryAnalysis(include=False)
+    cytoscape: CytoscapeAnalysis = CytoscapeAnalysis(include=False)
+    ml: MlAnalysis = MlAnalysis(include=False)
+    evaluation: EvaluationAnalysis = EvaluationAnalysis(include=False)
+
+    model_config = ConfigDict(extra='forbid')
+
+
+# The default length of the truncated hash used to identify parameter combinations
+DEFAULT_HASH_LENGTH = 7
+
+def label_validator(name: str):
+    """
+    A validator takes in a label
+    and ensures that it's alphanumeric (with underscores).
+    """
+    label_pattern = r'^\w+$'
+    def validate(label: str):
+        if not bool(re.match(label_pattern, label)):
+            raise ValueError(f"{name} label '{label}' contains invalid values. {name} labels can only contain letters, numbers, or underscores.")
+        return label
+    return validate
+
+class ContainerFramework(CaseInsensitiveEnum):
+    docker = 'docker'
+    # TODO: add apptainer variant once #260 gets merged
+    singularity = 'singularity'
+    dsub = 'dsub'
+
+class ContainerRegistry(BaseModel):
+    base_url: str
+    owner: str = Field(description="The owner or project of the registry")
+
+    model_config = ConfigDict(extra='forbid')
+
+class AlgorithmParams(BaseModel):
+    include: bool
+    directed: Optional[bool] = None
```

Comment:
what is this parameter?

Reply:
It's a holdout deprecated parameter ([see main branch code](https://github.com/Reed-CompBio/spras/blob/32a0424546a94b3d8f17362bcb4b6c4e6e6398fa/spras/config.py#L207-L209))

Reply:
Oh I see. I'm not sure if we should delete it. We want to make sure people who use old configs are still able to run the code which is why that part in the main branch code exists.

Reply:
I think we mentioned in an earlier conversation that we don't need deprecation logic at the moment. We can keep it here, though, since this PR should probably avoid as much breaking changes as possible.